design/aws-ecs-autoscaling-queue-worker.md

```diff
@@ -0,0 +1,225 @@
+
+## General approach
+
+The new [`ecs.QueueWorkerServiceBase`], [`ecs.Ec2QueueWorkerService`] and [`ecs.FargateQueueWorkerService`] classes will include L3 constructs for:
+
+* Ec2QueueWorkerService
+* FargateQueueWorkerService
+
+A `QueueWorkerService` will create a task definition with the specified container (on both EC2 and Fargate). An `SQSQueue` will be created and autoscaling of the ECS Service will be dependent on both CPU as well as the SQS queue's `ApproximateNumberOfMessagesVisible` metric. 
+
+## Code changes
+
+Given the above, we should make the following changes to support queue workers on ECS (for both EC2 and Fargate):
+1. Create `QueueWorkerServiceBaseProps` interface and `QueueWorkerServiceBase` construct
+2. Create `Ec2QueueWorkerServiceProps` interface  and `Ec2QueueWorkerService` construct
+3. Create `FargateQueueWorkerServiceProps` interface  and `FargateQueueWorkerService` construct
+
+# Part 1: Create `QueueWorkerServiceBaseProps` interface  and `QueueWorkerServiceBase` construct  
+
+The `QueueWorkerServiceBaseProps` interface will contain common properties used to construct both the Ec2QueueWorkerService and the FargateQueueWorkerService:
+
+```ts
+/**
+ * Properties to define an Query Worker service
+ */
+export interface QueueWorkerServiceBaseProps {
+  /**
+   * Cluster where service will be deployed
+   */
+  readonly cluster: ICluster;
+
+  /**
+   * The image to start.
+   */
+  readonly image: ContainerImage;
+
+  /**
+   * The CMD value to pass to the container. A string with commands delimited by commas.
+   *
+   * @default none
+   */
+  readonly command?: string;
+
+  /**
+   * Number of desired copies of running tasks
+   *
+   * @default 1
+   */
+  readonly desiredTaskCount?: number;
+
+  /**
+   * Flag to indicate whether to enable logging
+   *
+   * @default true
+   */
+  readonly enableLogging?: boolean;
+
+  /**
+   * The environment variables to pass to the container.
+   *
+   * @default none
+   */
+  readonly environment?: { [key: string]: string };
+
+  /**
+   * A name for the queue.
+   *
+   * If specified and this is a FIFO queue, must end in the string '.fifo'.
+   *
+   * @default CloudFormation-generated name
+   */
+  readonly queueName?: string;
+
+  /**
+   * Maximum capacity to scale to.
+   *
+   * @default (desiredTaskCount * 2)
+   */
+  readonly maxScalingCapacity?: number
+
+  /**
+   * The intervals for scaling.
+   *
+   * Maps a range of metric values to a particular scaling behavior.
```

Comment:
Could this be more clear? For someone perhaps not familiar with the autoscaling library, this may not be obvious.

Reply:
Updated and added a link to StepScaling public documentation.